forwards_paper.tex

```diff
@@ -196,14 +196,40 @@ \section*{Results}
 %%%%%%
 \subsection*{Simulation benchmarks}
 
-\plr{Jaime and Kevin to describe simulation methods here (separately).}
-
-Comparison of simulation with/without msprime, using \simupop{}
-or maybe just a simple haploid simulation with 1000 QTL and stabilizing selection on a trait (say).
-
-Maybe an estimate of how long \emph{just} the ARG recording and simplification takes,
-so that then we can say how fast the simulator would have to be to do $10^6$ whole chromosomes for $10^7$ generations
-in a day.
+To measure the performance gains from ancestry recording, we ran simulations implemented in C++ (see above) with and
+without ancestry tracking.  The simulations without ancestry tracking introduced neutral mutations at a scaled mutation
+rate equal to the scaled recombination rate (\textit{e.g.}, $\theta = 4N\mu = \rho$, where $\rho = 4Nr$, $N$ is the
+diploid population size, $\mu$ is the
+mutation rate to neutral mutations (per gamete, per generation) and $r$ is the recombination rate per diploid per
+generation.  Both $\mu$ and $r$ are the means of a Poisson process and we used the infinitely-many sites mutation model
+\citep{Kimura1969-uc}.
+
+We varied $N \in (1000, 10000, 50000)$, $\rho \in (10^3, 10^4, 10^5)$ and ran a single replicate of all $(N,\rho)$
+combinations.  Deleterious mutations were introduced at rate $\rho/100$ per generation, drawing scaled selection
+coefficients ($2Ns$)
+from a Gamma distribution with a mean of -5 and a shape parameter of 1.  This distribution of fitness effects results in
+thousands of weakly-deleterious mutations segregating in the population, many of which drift to intermediate
+frequencies.
+
+We ran all benchmarks on an Ubuntu Linux (version 16.04) system with two 2.6 Gigahertz Intel E5-2650 CPU with
+hyperthreading enabled.  We ran one simulation at a time and the machine was under minimal load otherwise.  We used GNU
+parallel \citep{Tange2011a} to kill
+any simulation that did not finish within 72 hours.  We used the Linux \texttt{time} command to record run time and peak
+memory usage of each replicate and our simulation scripts separately recorded the time spent in various steps of each
+run.
+
+For efficiency reasons, we only ran \simupop{} with ancestry tracking.
+
+\krt{Jaime--any details here to add re: how simuPOP was tweaked??}
```

Comment:
After brief discussion with @petrelharp , I'm thinking we should try to run a second set of sims for simuPOP. (We'd never revisited the netural muts after re-discovering alleletype=mutant!). We could do this with a wall time of 24 hr and cut out the largest parameters. Thoughts? (We can also discuss this on gchat, shortly)

Reply:
Take a look at first line of the results:

https://github.com/petrelharp/ftprime_ms/blob/master/sims/simupop_timings.txt#L2

That's 15 hours and 153Gb of RAM.

I'm fine running this, but I'm guessing we'll see all but 1 or 2 param combos being killed.  The fastest result was 1.33 hours, which was 318 times slower than the c++ implementation.

Reply:
Yeah ... it is slow. I don't want to waste too much kWh, so the question of whether to do this gets to what our point of including the simupop stuff is. Is it?

1. we can increase speed of simupop over tracking neutral arg?
2. simupop is slower than the c++ implementation?
3. we can run a simulation of X size in Y hours in simupop with this method + our glue code in ftprime?

If we want to do 1, it's worth running some neural sims for comparison. Maybe @petrelharp / @jeromekelleher can weigh in on whether it's worth it --- as it relates to the big picture here. If we think it is, but also don't want to waste Kevin's compute time too much, we could also just run the simupop benchmarks with smaller params. 

Regardless, I will set up my script to also tracking neutral variation and get back as to whether this will be doable.

As for 3, another wrinkle is that the change to binary data transfer does seem to have sped up ftprime significantly. So if we were to say re-run all the benchmarks on a grid of smaller params we may not waste as much time and compute as in the first run

Reply:
I think something should be done here.  As is, we haven't shown off ftprime itself yet, right?

We may simply have to do smaller simulations with simuPOP.  I am surprised that it turned out to be this much slower than fwdpy11.

What is this "binary data transfer" thing?

Reply:
I mentioned it briefly in an email thread on Wednesday, but didn't elaborate. 

Basically RecombCollector receives data from simuPOP on recombination/birth events. We had been getting it in plain text but I switched to binary. It seemed not to make a large % difference on small examples but I ran some large sims over the weekend in which is sped things up immensely

Reply:
OK--that is good news.  Seems like I should re-run the benchmarks with this change?